subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/Codecs.kt

```diff
@@ -119,103 +119,129 @@ class Codecs(
-        bind(ArtifactCollectionCodec(calculatedValueContainerFactory, artifactSetConverter))
-        bind(ImmutableAttributesCodec(attributesFactory, managedFactoryRegistry))
-        bind(AttributeContainerCodec(attributesFactory, managedFactoryRegistry))
-        bind(ComponentVariantIdentifierCodec)
-        bind(InitialTransformationNodeCodec(buildOperationExecutor, calculatedValueContainerFactory))
-        bind(ChainedTransformationNodeCodec(buildOperationExecutor, calculatedValueContainerFactory))
-        bind(TransformationStepCodec(inputFingerprinter))
-        bind(TransformationChainCodec())
-        bind(DefaultTransformerCodec(fileLookup, actionScheme))
-        bind(DefaultResolvableArtifactCodec(calculatedValueContainerFactory))
-        bind(TransformStepSpecCodec)
-        bind(PublishArtifactLocalArtifactMetadataCodec)
-        bind(TransformedProjectArtifactSetCodec())
-        bind(TransformedExternalArtifactSetCodec())
-        bind(CalculateArtifactsCodec(calculatedValueContainerFactory))
-        bind(TransformedArtifactCodec(calculatedValueContainerFactory))
-        bind(LocalFileDependencyBackedArtifactSetCodec(instantiator, attributesFactory, calculatedValueContainerFactory))
-        bind(CalculatedValueContainerCodec(calculatedValueContainerFactory))
-        bind(IsolateTransformerParametersNodeCodec(parameterScheme, isolatableFactory, buildOperationExecutor, classLoaderHierarchyHasher, fileCollectionFactory, documentationRegistry))
-        bind(FinalizeTransformDependenciesNodeCodec())
-        bind(ResolveArtifactNodeCodec)
-        bind(WorkNodeActionCodec)
-        bind(CapabilitySerializer())
-
-        bind(DefaultCopySpecCodec(patternSetFactory, fileCollectionFactory, instantiator))
-        bind(DestinationRootCopySpecCodec(fileResolver))
-
-        bind(TaskReferenceCodec)
-
-        bind(CachedEnvironmentStateCodec)
-
-        bind(IsolatedManagedValueCodec(managedFactoryRegistry))
-        bind(IsolatedImmutableManagedValueCodec(managedFactoryRegistry))
-        bind(IsolatedJavaSerializedValueSnapshotCodec)
-        bind(IsolatedArrayCodec)
-        bind(IsolatedSetCodec)
-        bind(IsolatedListCodec)
-        bind(IsolatedMapCodec)
-        bind(MapEntrySnapshotCodec)
-        bind(IsolatedEnumValueSnapshotCodec)
-        bind(StringValueSnapshotCodec)
-        bind(IntegerValueSnapshotCodec)
-        bind(FileValueSnapshotCodec)
-        bind(BooleanValueSnapshotCodec)
-        bind(NullValueSnapshotCodec)
-
-        bind(ServicesCodec())
-
-        bind(ProxyCodec)
-
-        // Java serialization integration
-        bind(unsupported<Externalizable>(NotYetImplementedJavaSerialization))
-        bind(JavaObjectSerializationCodec(javaSerializationEncodingLookup))
-
-        bind(BeanSpecCodec)
+    val userTypesBindings: Bindings
+    private
+    val fingerprintUserTypesBindings: Bindings
+
+    init {
+        fun makeUserTypeBindings(providersBlock: BindingsBuilder.() -> Unit) = Bindings.of {
```

Comment:
🎨 I think it would be cleaner to have this function in a companion object and use field initializers instead of `init`.

Reply:
I decided to go with the nested function, because it captures ~20 constructor arguments. Without it it either becomes 20-parameter function or we have to store all services passed to constructor in the `Codecs` instance, even though we only need them once.

Reply:
I see 😅